connmgr/connmanager.go

```diff
@@ -0,0 +1,336 @@
+type handleDisconnected struct {
+	id    uint64
+	retry bool
+}
+
+// handleFailed is used to remove a pending connection.
+type handleFailed struct {
+	c   *ConnReq
+	err error
+}
+
+// ConnManager provides a manager to handle network connections.
+type ConnManager struct {
+	// The following variables must only be used atomically.
+	stop         int32
+	connReqCount uint64
+
+	cfg            Config
+	wg             sync.WaitGroup
+	conns          map[uint64]*ConnReq
+	failedAttempts uint64
+	requests       chan interface{}
+	quit           chan struct{}
+}
+
+// handleFailedConn handles a connection failed due to a disconnect or any
+// other failure. If permanent, it retries the connection after the configured
+// retry duration. Otherwise, if required, it makes a new connection request.
+// After maxFailedConnectionAttempts new connections will be retried after the
+// configured retry duration.
+func (cm *ConnManager) handleFailedConn(c *ConnReq, retry bool) {
+	if atomic.LoadInt32(&cm.stop) != 0 {
+		return
+	}
+	if retry && c.Permanent {
+		c.retryCount++
+		d := time.Duration(c.retryCount) * cm.cfg.RetryDuration
+		if d > maxRetryDuration {
+			d = maxRetryDuration
+		}
+		log.Debugf("Retrying connection to %v in: %v", c.Addr, d)
+		time.AfterFunc(d, func() {
+			cm.Connect(c)
+		})
+	} else if cm.cfg.GetNewAddress != nil {
+		cm.failedAttempts++
+		if cm.failedAttempts > maxFailedAttempts {
+			log.Debugf("Max failed connection attempts reached: [%d] "+
+				"-- retrying connection in: %v", maxFailedAttempts,
+				cm.cfg.RetryDuration)
+			time.AfterFunc(cm.cfg.RetryDuration, func() {
+				cm.Connect(&ConnReq{Addr: cm.cfg.GetNewAddress()})
+			})
+		} else {
+			go cm.Connect(&ConnReq{Addr: cm.cfg.GetNewAddress()})
+		}
+	}
+}
+
+// connHandler handles all connection related requests.  It must be run as a
+// goroutine.
+//
+// The connection handler makes sure that we maintain a pool of active outbound
+// connections so that we remain connected to the network.  Connection requests
+// are processed and mapped by their assigned ids.
+func (cm *ConnManager) connHandler() {
+out:
+	for {
+		select {
+		case req := <-cm.requests:
+			switch msg := req.(type) {
+
+			case handleConnected:
+				connReq := msg.c
+				connReq.updateState(ConnEstablished)
+				connReq.conn = msg.conn
+				cm.conns[connReq.id] = connReq
+				log.Infof("Connection established: %v", connReq.Addr)
+				connReq.retryCount = 0
+				cm.failedAttempts = 0
+
+				if cm.cfg.OnConnection != nil {
+					go cm.cfg.OnConnection(connReq, msg.conn)
+				}
+
+			case handleDisconnected:
+				if connReq, ok := cm.conns[msg.id]; ok {
+					connReq.updateState(ConnDisconnected)
+					if connReq.conn != nil {
+						connReq.conn.Close()
+					}
+					log.Debugf("Connection closed: %v", connReq.Addr)
+					delete(cm.conns, msg.id)
+
+					if cm.cfg.OnDisconnection != nil {
+						go cm.cfg.OnDisconnection(connReq)
+					}
+
+					cm.handleFailedConn(connReq, msg.retry)
+				}
```

Comment:
I think an error log message should be added here in the scenario that we get a `handleDisconnected` with an id that we don't know of. At the moment, I can't think of a case where this would arise, but it may help us track down bugs in the future.